src/persist-client/src/internal/state.rs

```diff
@@ -2415,22 +2453,43 @@ where
         }
 
         let seqnos_since_last_rollup = self.seqno.0.saturating_sub(latest_rollup_seqno.0);
-        // every `threshold` seqnos since the latest rollup, assign rollup maintenance.
-        // we avoid assigning rollups to every seqno past the threshold to avoid handles
-        // racing / performing redundant work.
-        if seqnos_since_last_rollup > 0 && seqnos_since_last_rollup % u64::cast_from(threshold) == 0
-        {
-            return Some(self.seqno);
-        }
 
-        // however, since maintenance is best-effort and could fail, do assign rollup
-        // work to every seqno after a fallback threshold to ensure one is written.
-        if seqnos_since_last_rollup
-            > u64::cast_from(
-                threshold * PersistConfig::DEFAULT_FALLBACK_ROLLUP_THRESHOLD_MULTIPLIER,
-            )
-        {
-            return Some(self.seqno);
+        if use_active_rollup {
+            // If sequnos_since_last_rollup>threshold, and there is no existing rollup in progress,
+            // we should start a new rollup.
+            if seqnos_since_last_rollup > u64::cast_from(threshold)
+                && self.active_rollup().is_none()
+            {
+                return Some(self.seqno);
+            }
+
+            // If we have an active rollup, we need to check if it has been running too long.
+            // If it has, we should start a new rollup.
+            // This is to guard against a worker dying/taking too long/etc.
```

Comment:
This has slightly surprising-to-me behaviour: if there's a old/timed-out rollup here, we'll kick off a new one at the current seqno even if the most recent rollup is recent enough for us.

Should only really come up if the threshold is being tuned, though, which should be rare. Happy to keep it as is if you find this behaviour more intuitive.

Reply:
Good point. I'm fairly certain it shouldn't be possible to arrive in such a situation with my understanding of the state machine here, but I will move the logic into the guard to make it clearer.